Report web-apps/formplayer request failures to datadog & sentry

dannyroberts opened the pull request:
##### SUMMARY

Our current web apps / formplayer metrics/logging strategy relies on (1) formplayer returning non-2xx responses for any kind of error and (2) formplayer reporting any issues it has to datadog/sentry. I think both of those are provably untrue, and not super-quickly fixable, so I think we need corroborating metrics that we report to HQ from the browser when we encounter clear/known errors states.

This PR reports to datadog and sentry whenever a request to formplayer fails (for any reason) while filling out a form. To datadog we report the values that have a fixed set of values like status code, status text (which I think is "accepted"/"rejected"), and also domain. To sentry we report everything we've got.

dannyroberts committed:
Report web-apps/formplayer request failures to datadog & sentry

dannyroberts committed:
Respond to stickler

dannyroberts commented on corehq/apps/cloudcare/static/cloudcare/js/form_entry/webformsession.js:
```diff
@@ -534,3 +538,30 @@ WebFormSession.prototype.renderFormXml = function (resp, $form) {
+        contentType: "application/json",
+        dataType: "json",
+        success: function () {
+            console.info('Successfully reported error: ' + JSON.stringify(data));
```
I think this is okay

dannyroberts commented on corehq/apps/cloudcare/static/cloudcare/js/form_entry/webformsession.js:
```diff
@@ -534,3 +538,30 @@ WebFormSession.prototype.renderFormXml = function (resp, $form) {
+            console.info('Successfully reported error: ' + JSON.stringify(data));
+        },
+        error: function () {
+            console.error('Failed to report error: ' + JSON.stringify(data));
```
I think this is okay

dannyroberts committed:
Include report_formplayer_error url in app review as well

dannyroberts committed:
Prevent an error in reportFormplayerErrorToHQ from crashing the page

orangejenny commented on corehq/apps/cloudcare/static/cloudcare/js/form_entry/webformsession.js:
```diff
@@ -534,3 +538,30 @@ WebFormSession.prototype.renderFormXml = function (resp, $form) {
+        statusText: resp.statusText,
+        state: resp.state ? resp.state() : null,
+        status: resp.status,
+        domain: self.domain,
```
Nitpick: is domain needed, given that it's also embedded in the URL?

dannyroberts replied:
Maybe not, but my thought was that it doesn't hurt. This will make it show up in the sentry "additional data" section, which may or may not be useful given it's other places too

orangejenny commented on corehq/apps/cloudcare/views.py:
```diff
@@ -461,3 +463,17 @@ def put(self, request, *args, **kwargs):
+        'status': data.get('status'),
+        'domain': domain,
+    })
+    notify_error(message='Formplayer error reported from the browser', details=data)
```
Can this go to the formplayer project in sentry?

dannyroberts replied:
Oh, hmm. I think there's not a simple way to do that from HQ. SInce it's coming through HQ and not reported directly from formplayer, I think it's fine for it to show up in the commcarehq sentry project though

orangejenny reviewed (approved):
Looks reasonable. Will this affect all formplayer errors, or only those that are not currently sent to sentry?

dannyroberts committed:
Report any web apps error messages we show to the user to datadog & sentry

dannyroberts commented on corehq/apps/cloudcare/static/cloudcare/js/form_entry/webformsession.js:
```diff
@@ -534,3 +538,30 @@ WebFormSession.prototype.renderFormXml = function (resp, $form) {
+    $.ajax({
+        type: 'POST',
+        url: self.urls.errorReportUrl,
+        data: JSON.stringify(data),
```
Yeah if you do POST with `data: {...}` it'll format that the properties as post parameters (a=1&b=2). If you send a string there instead, it uses the contents of the string verbatim, in this case to make the POST body a JSON string.

It's possible that specifying `dataType: "json",` here changes the behavior, I'm not sure. So I should say if it has any affect at all, it's what I said above, but I didn't try it the other way.

dannyroberts commented on corehq/apps/cloudcare/static/cloudcare/js/form_entry/webformsession.js:
```diff
@@ -534,3 +538,30 @@ WebFormSession.prototype.renderFormXml = function (resp, $form) {
+        action: action,
+        readableErrorMessage: errorMessage,
+        statusText: resp.statusText,
+        state: resp.state ? resp.state() : null,
```
I think this is the meta status of the http request—like whether it was able to connect with the server at all. I've seen the values "accepted" and "rejected" here and I think nothing else so far

dannyroberts committed:
Respond to stickler

dannyroberts committed:
Simplify exposing reportFormplayerErrorToHQ in webformsession.js

orangejenny commented on corehq/apps/cloudcare/views.py:
```diff
@@ -468,12 +470,32 @@ def put(self, request, *args, **kwargs):
+
+
+def _message_to_tag_value(message, allowed_chars=string.ascii_lowercase + '_'):
+    message_tag = '_'.join(unidecode(message).split(' ')[:4]).lower()
```
I'm not following the `:4`, what's that?

dannyroberts replied:
Yeah sorry not the easiest to follow.

Added a docstring with examples to help make it clearer

orangejenny replied:
thanks!

dannyroberts committed:
Add docstring to _message_to_tag_value

millerdev commented on corehq/apps/cloudcare/static/cloudcare/js/util.js:
```diff
@@ -121,6 +130,28 @@ hqDefine('cloudcare/js/util', function () {
+                contentType: "application/json",
+                dataType: "json",
+                success: function () {
+                    console.info('Successfully reported error: ' + JSON.stringify(data));
```
I think you can use `window.console` to avoid stickler comments (when you really want console logging in committed code).

dannyroberts replied:
🙇 https://github.com/dimagi/commcare-hq/pull/27412/commits/c89870ad59ceb26096c97818bf8b293bae2f8149

millerdev commented on corehq/apps/cloudcare/static/cloudcare/js/util.js:
```diff
@@ -121,6 +130,28 @@ hqDefine('cloudcare/js/util', function () {
+                },
+            });
+        } catch (e) {
+            console.error("reportFormplayerErrorToHQ failed hard and there is no where else to report this error", e);
```
Maybe include `data` in this error message? It seems unlikely that the `error` callback will fire if there is an exception caught here.
```suggestion
            console.error(
                "reportFormplayerErrorToHQ failed hard and there is nowhere " +
                "else to report this error: " + JSON.stringify(data),
                e,
            );
```

dannyroberts committed:
Change commcare.formplayer.webformsession_request_failure to use "request" tag

(instead of "action") to match the name for formplayer-originated metrics
that we already have.

dannyroberts committed:
Show error report contents in failure to report error

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

dannyroberts committed:
Appease stickler with console => window.console

dannyroberts committed:
Remove trailing comma from js function arg list

dannyroberts committed:
Have _message_to_tag_value allow digits

dannyroberts commented:
Build 3 timed out, I'm going to call that a "pass". Pushed another commit that I ran tests for and am OK not running full tests on to get this out the door

ShivamPokhriyal reviewed (commented):
Looks, great! Have one doubt. In case Formplayer throws an exception like `FormNotFound` do we wanna log that as well from HQ? These exceptions gets logged to Sentry btw afaik.

dannyroberts commented on corehq/apps/cloudcare/static/cloudcare/js/form_entry/webformsession.js:
```diff
@@ -251,6 +252,19 @@ WebFormSession.prototype.handleFailure = function (resp, action, textStatus, fai
+    hqImport('cloudcare/js/util').reportFormplayerErrorToHQ({
+        type: 'webformsession_request_failure',
+        request: action,
+        readableErrorMessage: errorMessage,
```
Hey Shivam, this is a good question and something I'm not 100% sure about. But when I was working on this (and I double checked just now), I found that a "success" response (200 response) could not cause the spinner next to the answer to turn into a ❌or make the error text show up beneath the answer. It can show a red banner error, but I'm already logging those separately from this, and those are showing up there.
